### ISSUE TITLE
Time-series Line / Bar chart: Add option to plot 0 if no data.

### ISSUE THREAD

**Issue by EBoisseauSierra** [apache/superset#15036]
## Issue

I have a simple metric (say: `SUM(orders_count)`), and I would like to track its evolution over time. My objective is to visualize the total for different time granularity (group by hour, day, week, etc.).

However, I don't always have records for each time bucket (e.g. I didn't had any order between 2 and 3am, or on Sunday, etc.). In this situation, **I want to plot `0` for each time bucket I had no data for**.

In that situation, the line chart simply omit the given time bucket and interpolate the line given the previous and next data point:

![Screenshot from 2021-06-08 09-26-46_shadow](https://user-images.githubusercontent.com/37387755/121152480-27218300-c83d-11eb-8394-798110b73378.png)

I am aware that I can use `pandas` resampling methods to actually plot a `0` data point on each given hour I had no order:

![Screenshot from 2021-06-08 09-27-10_shadow](https://user-images.githubusercontent.com/37387755/121153236-cc3c5b80-c83d-11eb-8b04-e74029efad46.png)

However, this workaround doesn't “follow” the time granularity I aggregate data on:

![Screenshot from 2021-06-08 09-27-37_shadow](https://user-images.githubusercontent.com/37387755/121154367-d6128e80-c83e-11eb-85e7-000374fe2daf.png)

This means that each time I want to update the granularity, I have to modify it at two different places to get the correct graph. 

![Screenshot from 2021-06-08 09-27-55_shadow](https://user-images.githubusercontent.com/37387755/121155016-5802b780-c83f-11eb-9243-b9b1d74fc1cf.png)

Moreover, this makes that I cannot pass the time granularity as a parameter from a native filter.

## Requested feature

I would like a simple option to replace missing values (i.e. time buckets with no record to aggregate) with either:

* linear interpolation (current behaviour),
* nothing (can be emulated via `pandas.resample(<granularity>, mean)`), 
* 0 (can be emulated via `pandas.resample(<granularity>, sum)`).

Metabase uses a simple dropdown menu for that:

![Screenshot from 2021-06-08 11-02-27_shadow.png](https://user-images.githubusercontent.com/37387755/121167244-a917a900-c849-11eb-9086-7d76927e2a54.png)

## Alternatives

As seen above, using `pandas.resample` is solving the issue partially only, as it doesn't dynamically adjust to the selected time granularity.

One could of course write a custom query that joins the list of every single hour between `min(timestamp)` and `max(timestamp)` to force generate records for these time buckets… but it's a lot of work — and again wouldn't be dynamically adapting to different time grains.

## Context

Examples generated on Superset 1.1.0.

**Comment by EBoisseauSierra** [comment]
Very much related to following PRs:

* #3512 Making data gaps visible in line charts by inserting zeros (closed in 2017),
* #487 Option for Timeseries Null days to be shown (closed in 2018)

**Comment by zhaoyongjie** [comment]
related issue: https://github.com/apache/superset/issues/14365

**Comment by NickChoiRBX** [comment]
For my purposes, it'd be much better if there was a linear interpolation option. Is there a version that's available on? I've seen a lot of people requesting this be made available as a feature, given that the app's UI used to behave this way.

**Comment by EBoisseauSierra** [comment]
@NickChoiRBX Isn't linear interpolation the current behaviour already (see first screenshot)?

**Comment by NickChoiRBX** [comment]
![image](https://user-images.githubusercontent.com/68253528/122974361-10a01f00-d347-11eb-97bf-5d1f4ba23182.png)
I've seen mostly people mentioning that this functionality is no longer available for them. I'm on version 1.0.1 and this is how line charts look for me (notice the point in the upper left corner for the purple line).

**Comment by rusackas** [comment]
I know that @zhaoyongjie is working on advanced analytics (including resampling, I believe) for the newer Timeseries Chart. I hope that this helps resolve things for you soon.

I'm _not_ sure, offhand, if it would be easy to add an option to the resample controls that sets grain to "Inhert Time Grain" or something to that effect. Then you wouldn't have to double-configure it.

It _may_ also be possible to add a control to the chart which works a bit like the [PR you found](https://github.com/apache/superset/pull/3512/files) to fill in the data (based on the granularity) with zeros. Or better, yet, make it a choice to "infill data" and a secondary choice to fill with `0` points or `null` for anyone who doesn't like the connecting line. My _suspicion_ is that most people want interpolation between _known_ datapoints, but the second-most-common request would be the `null`s inserted, to create gaps. `0` points doesn't seem like a terribly common need, but could be supported by these controls, I think.

Curious what @villebro thinks here too... about the general approach, as well as the dynamic/progressive-reveal controls that might be needed for my half-baked proposal ;)

**Comment by zhaoyongjie** [comment]
Thanks for the proposal, I will finish it in the resample(advanced analytics).

**Comment by EBoisseauSierra** [comment]
To provide some context: We believe “0-filling” (in addition to interpolation and gaps) can be relevant when, e.g., you want to plot the count of records or sum of a given metric per <time_grain>. In such cases, not having data (for a given time bucket) means something — i.e., that the count/sum *is actually 0*.

**Comment by rusackas** [comment]
Yep, I've definitely had needs for 0-filling and null-filling (gaps) in the past, so I'd love to find a good design for controls to allow any of these options.
